src/Strategy/Swift/PackageSwift.hs

```diff
@@ -0,0 +1,235 @@
+import Path
+import Text.Megaparsec (
+  MonadParsec (takeWhile1P, try),
+  Parsec,
+  anySingle,
+  between,
+  empty,
+  sepEndBy,
+  skipManyTill,
+ )
+import Text.Megaparsec.Char (space1)
+import Text.Megaparsec.Char.Lexer qualified as Lexer
+
+-- | Parsing
+-- *
+type Parser = Parsec Void Text
+
+sc :: Parser ()
+sc =
+  Lexer.space
+    space1
+    (Lexer.skipLineComment "//")
+    (Lexer.skipBlockComment "/*" "*/")
+
+lexeme :: Parser a -> Parser a
+lexeme = Lexer.lexeme sc
+
+symbol :: Text -> Parser Text
+symbol = Lexer.symbol sc
+
+scWOComment :: Parser ()
+scWOComment = Lexer.space space1 empty empty
+
+symbolWOComment :: Text -> Parser Text
+symbolWOComment = Lexer.symbol scWOComment
+
+betweenDoubleQuotes :: Parser a -> Parser a
+betweenDoubleQuotes = between (symbol "\"") (symbol "\"")
+
+betweenSquareBrackets :: Parser a -> Parser a
+betweenSquareBrackets = between (symbol "[") (symbol "]")
+
+betweenBrackets :: Parser a -> Parser a
+betweenBrackets = between (symbol "(") (symbol ")")
+
+maybeComma :: Parser ()
+maybeComma = void $ optional $ lexeme $ symbol ","
+
+parseQuotedText :: Parser Text
+parseQuotedText = betweenDoubleQuotes (lexeme $ takeWhile1P (Just "quoted text") (/= '"'))
+
+parseKeyValue :: Text -> Parser a -> Parser a
+parseKeyValue t parser = lexeme $ symbol (t <> ":") *> parser
+
+isEndLine :: Char -> Bool
+isEndLine '\n' = True
+isEndLine '\r' = True
+isEndLine _ = False
+
+-- | Represents https://github.com/apple/swift-package-manager/blob/main/Documentation/PackageDescription.md#methods.
+data SwiftPackage = SwiftPackage
+  { swiftToolVersion :: Text
+  , packageDependencies :: [SwiftPackageDep]
+  }
+  deriving (Show, Eq, Ord)
+
+-- | Represents https://github.com/apple/swift-package-manager/blob/main/Documentation/PackageDescription.md#package-dependency.
+data SwiftPackageDep
+  = GitSource SwiftPackageGitDep
+  | PathSource Text
+  deriving (Show, Eq, Ord)
+
+data SwiftPackageGitDep = SwiftPackageGitDep
+  { srcOf :: Text
+  , versionRequirement :: Maybe SwiftPackageGitDepRequirement
+  }
+  deriving (Show, Eq, Ord)
+
+data SwiftPackageGitDepRequirement
+  = Branch Text
+  | Revision Text
+  | Exact Text
+  | From Text
+  | UpToNextMajor Text
+  | UpToNextMinor Text
+  | ClosedInterval (Text, Text)
+  | RhsHalfOpenInterval (Text, Text)
+  deriving (Show, Eq, Ord)
+
+toConstraint :: SwiftPackageGitDepRequirement -> VerConstraint
+toConstraint (Branch b) = CEq b
+toConstraint (Revision r) = CEq r
+toConstraint (Exact e) = CEq e
+-- from constraint is equivalent to upToNextMajor
+-- Reference: https://github.com/apple/swift-package-manager/blob/main/Documentation/PackageDescription.md#methods-3
+toConstraint (From f) = CEq $ "^" <> f
+toConstraint (UpToNextMajor c) = CEq $ "^" <> c
+toConstraint (UpToNextMinor c) = CEq $ "~" <> c
+toConstraint (ClosedInterval (lhs, rhs)) = CEq $ ">=" <> lhs <> " " <> "<=" <> rhs
+toConstraint (RhsHalfOpenInterval (lhs, rhs)) = CEq $ ">=" <> lhs <> " " <> "<" <> rhs
```

Comment:
My preference is we go with (2) leave it way it is because it is more correct. 

If (1) is followed, we might fail to resolve versions when LHS or RHS of constraint does not exist (not a perfect example but ~ e.g. >= 0.0.0, where 0.0.0 does not exist)

I've validated both `RhsHalfOpenInterval`, `ClosedInterval`, and others in unit and integration test for swift fetchers:

- (integration test in fetcher) https://github.com/fossas/FOSSA/pull/6418/files#diff-afd29de33fe5b6b70a89eaaafe634ffae1c7a94d3d081ed77522295b625a942dR83
- (spectrometer unit test) - https://github.com/fossas/spectrometer/pull/354/files#diff-68e91f6fa560fd2d45ebf74cd2eacc278a5fff0d66e2a7d218e9480661a5ae48R71

Reply:
I'm comfortable with option 2 as well, mainly because of our conversation yesterday about how the Swift fetcher is built to handle these constraints.

Could we actually add a type similar to `CRhsHalfOpenInterval` and others for this PR? I think that would ease all of my concerns about version handling and it shouldn't be a hard addition. We're going to need it eventually and I think this `VersionConstraint` would make sure we always handle this version case the same everywhere.